server/repository/src/lib.rs

```diff
@@ -37,3 +38,51 @@ pub fn run_db_migrations(connection: &StorageConnection) -> Result<(), String> {
     );
     Ok(())
 }
+
+use std::fmt::Debug as DebugTrait;
+pub trait Delete: DebugTrait {
+    fn delete(&self, con: &StorageConnection) -> Result<(), RepositoryError>;
+    // Test only
+    fn assert_deleted(&self, con: &StorageConnection);
+}
+
+pub trait Upsert: DebugTrait {
+    fn upsert_sync(&self, con: &StorageConnection) -> Result<(), RepositoryError>;
+    // Test only
+    fn assert_upserted(&self, con: &StorageConnection);
+    // Test only, can be used to drill down to concrete type (see test below)
+    // also casting to any must be implemented by concrete type to be able to downcast to it
+    // This is needed for integration test (where test record is generated for inventory adjustment, but id is not know until site is created)
+    fn as_mut_any(&mut self) -> Option<&mut dyn Any> {
```

Comment:
Hmm, I did weight those two options, but didn't spend too long, on it. I think the thought process was: We use the same record type for tests and for integration:

https://github.com/msupply-foundation/open-msupply/blob/bc83acbb3cd51fe3375ac01135c6eb880a793936/server/service/src/sync/test/mod.rs#L12

https://github.com/msupply-foundation/open-msupply/blob/bc83acbb3cd51fe3375ac01135c6eb880a793936/server/service/src/sync/translations/mod.rs#L150

https://github.com/msupply-foundation/open-msupply/blob/bc83acbb3cd51fe3375ac01135c6eb880a793936/server/service/src/sync/translations/mod.rs#L127-L130

And therefore Upsert will always be together with UpsertTest, and at that point i stopped and just went with one combined trait. 

Does that sound ok for now ?

Reply:
Yes sure. Probably a bit cleaner to have it separated(?) but also more work to have two implementations...

Reply:
I think it would be messier to have them separated, because then the IntegrationOperation would need two traits requirement, and it would be harder to explain to the future implementers, which trait is used where, and which trait definitions are required (unless maybe Upsert trait had the test trait as pre-requisite)